pkg/metrics/metrics.go

```diff
@@ -105,46 +171,95 @@ func NewMetricsManager() MetricsManager {
+func (opMgr *operationMetricsManager) OperationStart(key OperationKey, val OperationValue) {
+	val.startTime = time.Now()
+	opMgr.cache.LoadOrStore(key, val)
 }
 
-func (opMgr *operationMetricsManager) DropOperation(op Operation) {
+// OperationStart drops an operation
+func (opMgr *operationMetricsManager) DropOperation(op OperationKey) {
 	opMgr.cache.Delete(op)
 }
 
-func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status OperationStatus) {
-	obj, exists := opMgr.cache.Load(op)
+// RecordMetrics emits operation metrics
+func (opMgr *operationMetricsManager) RecordMetrics(opKey OperationKey, opStatus OperationStatus, snapshot *crdv1.VolumeSnapshot, driverName string) {
+	obj, exists := opMgr.cache.Load(opKey)
 	if !exists {
 		// the operation has not been cached, return directly
 		return
 	}
-	ts, ok := obj.(time.Time)
+	opVal, ok := obj.(OperationValue)
 	if !ok {
-		// the cached item is not a time.Time, should NEVER happen, clean and return
-		klog.Errorf("Invalid cache entry for key %v", op)
-		opMgr.cache.Delete(op)
+		// the cached item is not a OperationValue, should NEVER happen, clean and return
+		klog.Errorf("Invalid cache entry for key %v", opKey)
+		opMgr.cache.Delete(opKey)
 		return
 	}
-	strStatus := opStatusUnknown
-	if status != nil {
-		strStatus = status.String()
+	status := string(SnapshotStatusTypeUnknown)
+	if opStatus != nil {
+		status = opStatus.String()
 	}
-	duration := time.Since(ts).Seconds()
-	opMgr.opLatencyMetrics.WithLabelValues(op.Driver, op.Name, strStatus).Observe(duration)
-	opMgr.cache.Delete(op)
+
+	// if we do not know the driverName while recording metrics,
+	// refer to the cached version instead.
+	if driverName == "" || driverName == unknownDriverName {
+		driverName = opVal.Driver
+	}
+
+	operationDuration := time.Since(opVal.startTime).Seconds()
+	opMgr.opLatencyMetrics.WithLabelValues(driverName, opKey.Name, opVal.SnapshotType, status).Observe(operationDuration)
+
+	// Report cancel metrics if we are deleting an unfinished VolumeSnapshot
+	if opKey.Name == DeleteSnapshotOperationName {
+		// check if we have a CreateSnapshot operation pending for this
+		createKey := NewOperationKey(CreateSnapshotOperationName, opKey.ResourceID)
+		obj, exists := opMgr.cache.Load(createKey)
+		if exists {
+			// record a cancel metric if found
+			opMgr.recordCancelMetric(obj, createKey, operationDuration)
+		}
+
+		// check if we have a CreateSnapshotAndReady operation pending for this
+		createAndReadyKey := NewOperationKey(CreateSnapshotAndReadyOperationName, opKey.ResourceID)
+		obj, exists = opMgr.cache.Load(createAndReadyKey)
+		if exists {
+			// record a cancel metric if found
+			opMgr.recordCancelMetric(obj, createAndReadyKey, operationDuration)
+		}
+	}
+
+	opMgr.cache.Delete(opKey)
+}
+
+// recordCancelMetric records a metric for a create operation that hasn't finished
+func (opMgr *operationMetricsManager) recordCancelMetric(obj interface{}, key OperationKey, duration float64) {
```

Comment:
I think we should simply have "opMgr.opLatencyMetrics.WithLabelValues(
		val.Driver,
		key.Name,
		val.SnapshotType,
		string(SnapshotStatusTypeCancel)," in the RecordMetrics call. Cache management will happen and after this call again?

Reply:
if we're deleting this snapshot, won't that cause a leak?

When we're record a delete metric, we must
1. Load Delete operation key, record metric, and delete key from cache
2. Load Create operation key, record cancel metric, and delete key from cache
3. Load CreateAndReady operation key, record metric, and delete key from cache.

Since the operation name is part of the key, we have three unique keys to delete (if Create/CreateAndReady keys still exist)

Reply:
oh, yeah you are right. I read it wrongly :)